Tools/Scripts/webkitpy/api_tests/runner.py

```diff
@@ -155,53 +173,133 @@ def run(self, tests, num_workers):
+            self._num_workers = min(workers, max(effective_work_count, 1))
+            _log.info(f'Worker allocation: {len(shards)} shards + {current_test_parallel_safety_count} test-parallel-safety (batch 1/{len(test_parallel_safety_batches) if test_parallel_safety_batches else 0}) = {effective_work_count} workers needed, spawning {self._num_workers} (limit: {workers})')
+
             system_shards = {}
             non_system_shards = {}
 
             for name, tests in iteritems(shards):
                 group = self.port.group_for_shard(type('Shard', (), {'name': name})(), suite='api-tests')
-                if group == 'system' and not self.port.get_option('fully_parallel'):
+                if group == 'system' and not self.port.get_option('fully_parallel') and not test_parallel_safety_tasks:
                     system_shards[name] = tests
                 else:
                     non_system_shards[name] = tests
 
-            # Run non-system tests first
-            if non_system_shards:
-                non_system_groups = [group for group in mutually_exclusive_groups if group != 'system']
-                with TaskPool(
-                    workers=self._num_workers,
-                    mutually_exclusive_groups=non_system_groups,
-                    setup=setup_shard, setupkwargs=dict(port=self.port, devices=devices, log_limit=self.log_limit), teardown=teardown_shard,
-                ) as pool:
-                    was_sent = set()
-
-                    # Dispatch shards from non-system groups first
-                    for name, tests in iteritems(non_system_shards):
-                        group = self.port.group_for_shard(type('Shard', (), {'name': name})(), suite='api-tests')
-                        if group and group != 'system' and not self.port.get_option('fully_parallel'):
-                            was_sent.add(name)
-                            pool.do(run_shard, name, *tests, group=group)
-
-                    # Dispatch remaining non-system shards
-                    for name, tests in iteritems(non_system_shards):
-                        if name in was_sent:
-                            continue
-                        pool.do(run_shard, name, *tests)
-
-                    pool.wait()
-
-            # Run system tests after all non-system tests complete
-            if system_shards:
-                with TaskPool(
-                    workers=1,  # System tests run with single worker to avoid conflicts
-                    mutually_exclusive_groups=[],
-                    setup=setup_shard, setupkwargs=dict(port=self.port, devices=devices, log_limit=self.log_limit), teardown=teardown_shard,
-                ) as pool:
-                    for name, tests in iteritems(system_shards):
-                        pool.do(run_shard, name, *tests)
-
-                    pool.wait()
+            if non_system_shards or test_parallel_safety_batches:
```

Comment:
Seems like we should be pushing this logic earlier in the function. If we are adding new tests to validate in parallel, but we don't have any non-system shards, isn't this whole function useless?